pkg/netns/netns_linux.go

```diff
@@ -0,0 +1,200 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package netns
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"runtime"
+	"sync"
+
+	"golang.org/x/sys/unix"
+)
+
+type NetNSHandle struct {
+	fd int
+}
+
+// CreateNewNetns creates a network namespace and
+// returns a handle to it.
+//
+// The current thread will not be moved to this new network namespace as with a basic
+// UNSHARE syscall - the original network namespace will be retained and a handle to the
+// new network namespace will be returned.
+//
+// The namespace created by this call is NOT pinned, or bind-mounted, and so will be
+// closed/lost when the process terminates.
+//
+// If you need a pinned network namespace, call `CreateNewNetns` with pin == true, or call Pin() on
+// the returned handle.
+func New() (NetNSHandle, error) {
+	var innerError error
+	var newNetns NetNSHandle
+	restore, err := saveRestoreCurrentThreadNetns()
+	if err != nil {
+		return newNetns, err
+	}
+	defer restore()
+
+	// this invokes the UNSHARE syscall, which creates a new netns *and moves the current thread*
+	// to that netns. So, after creating and getting a FD, we need to move it BACK (via defer), so that
+	// it looks like we've never left the original netns.
+	fd, err := createAndMoveToNewNetNS()
+	if err != nil {
+		return NetNSHandle{fd: -1}, fmt.Errorf("Unable to create new netns: %w", err)
+	}
+	return NetNSHandle{fd: int(fd)}, innerError
+}
+
+// DoByPinPath is for when a process (CNI plugins, etc) needs to jump into a netns
+// managed/owned by a different process in order to do things, and that other process
+// has "pinned" or bind-mounted the netns to a file.
+//
+// Cilium code should not pin or bind-mount network namespaces it creates, this function is
+// purely for interacting with network namespaces created and managed outside of Cilium.
+func DoByPinPath(netNSPath string, runInside func(*NetNSHandle) error) error {
```

Comment:
In general, aside from the CNI which _must_ handle pinned NSes created elsewhere, the goal here was to keep this as a one-off to discourage people in the Cilium codebase from leveraging this API on external pinned NS paths.

It is more composable to split them, at the cost of depending on the caller to explicitly close to avoid a FD leak.

But given the other calls are now also making that assumption, agree it makes sense to split.